File: js/any_output_switch.js
/**
 * Any Output Switch (misc) — フロントエンド拡張
 *
 * 入力1本から、LATENT_01 / any_01 などの可変出力へ配線します。
 * 最後尾は常に index（INT）で、Python が選んだ出力番号を受け取れます。
 */

import { app } from "../../scripts/app.js";
import {
    INDEX_OUTPUT_NAME,
    IoDirection,
    SELECT_INDEX_DEFAULT,
    SELECT_INDEX_KEY,
    addDataOutputBeforeIndex,
    countDataOutputs,
    debounce,
    ensureIndexOutput,
    ensureMinimumDataOutputs,
    followConnectionUntilType,
    getDataOutputNamePrefix,
    getIndexOutputSlotIndex,
    isPrimitivesLinkType,
    slotLabelForLinkType,
    syncPrimitivesLinkSlot,
    propagatePrimitiveSplitSync,
    removeUnusedDataOutputsFromEnd,
    renumberDataOutputs,
    syncSelectIndexWidget,
} from "./utils.js";

/** Python の get_name("Any Output Switch") と同じ文字列 */
const NODE_CLASS = "Any Output Switch (misc)";

/** 常に表示するデータ出力の最小本数（any_01 の1本 + index） */
const MIN_DATA_OUTPUTS = 1;

/**
 * @param {object} node
 * @returns {object|undefined}
 */
function getMainInput(node) {
    return node.inputs?.find((inp) => inp.name === "input");
}

/**
 * @param {object} node
 * @returns {string}
 */
function resolveNodeType(node) {
    const mainInput = getMainInput(node);
    const mainIndex = mainInput ? node.inputs.indexOf(mainInput) : 0;

    // まず input 側の実配線型を見て、未確定なら出力側の配線から逆算する。
    let connectedType = followConnectionUntilType(node, IoDirection.INPUT, mainIndex, true);
    if (!connectedType?.type || connectedType.type === "*") {
        for (let i = 0; i < countDataOutputs(node); i++) {
            connectedType = followConnectionUntilType(node, IoDirection.OUTPUT, i, true);
            if (connectedType?.type && connectedType.type !== "*") {
                break;
            }
        }
    }
    return connectedType?.type || "*";
}

/**
 * @param {Function} nodeType
 */
function setupMiscAnyOutputSwitch(nodeType) {
    const onNodeCreated = nodeType.prototype.onNodeCreated;
    const onConnectionsChange = nodeType.prototype.onConnectionsChange;
    const onConfigure = nodeType.prototype.onConfigure;

    nodeType.prototype.serialize_widgets = true;

    nodeType.prototype.onNodeCreated = function () {
        const result = onNodeCreated?.apply(this, arguments);
        this.stabilizeBound = this.stabilize.bind(this);
        this.nodeType = "*";
        this.properties = this.properties ?? {};
        if (this.properties.select_index == null) {
            this.properties.select_index = SELECT_INDEX_DEFAULT;
        }

        const prefix = getDataOutputNamePrefix(this.nodeType);
        ensureMinimumDataOutputs(this, MIN_DATA_OUTPUTS, prefix, this.nodeType);
        this.stabilize();
        return result;
    };

    nodeType.prototype.onConfigure = function () {
        const result = onConfigure?.apply(this, arguments);
        requestAnimationFrame(() => this.scheduleStabilize(0));
        return result;
    };

    nodeType.prototype.onConnectionsChange = function (type, slotIndex, isConnected, linkInfo, ioSlot) {
        onConnectionsChange?.call(this, type, slotIndex, isConnected, linkInfo, ioSlot);
        if (app.configuringGraph) {
            return;
        }
        this.scheduleStabilize();
    };

    nodeType.prototype.onConnectionsChainChange = function () {
        if (app.configuringGraph) {
            return;
        }
        this.scheduleStabilize();
    };

    nodeType.prototype.scheduleStabilize = function (ms = 64) {
        return debounce(this.stabilizeBound, ms);
    };

    nodeType.prototype.stabilize = function () {
        if (this.removed) {
            return;
        }

        this.nodeType = resolveNodeType(this);
        const prefix = getDataOutputNamePrefix(this.nodeType);

        const mainInput = getMainInput(this);
        if (mainInput && isPrimitivesLinkType(this.nodeType)) {
            syncPrimitivesLinkSlot(mainInput, this.nodeType);
        } else if (mainInput) {
            mainInput.type = this.nodeType;
        }

        for (const input of this.inputs ?? []) {
            if (input.name === SELECT_INDEX_KEY) {
                input.type = "INT";
            }
        }

        ensureIndexOutput(this);
        removeUnusedDataOutputsFromEnd(this, MIN_DATA_OUTPUTS, prefix);
        ensureMinimumDataOutputs(this, MIN_DATA_OUTPUTS, prefix, this.nodeType);

        // 最後のデータ出力が使用中なら 1 本増やして常に「次の差し込み口」を確保する。
        const lastDataPos = countDataOutputs(this) - 1;
        if (lastDataPos >= 0 && this.outputs[lastDataPos]?.links?.length) {
            addDataOutputBeforeIndex(this, 1, prefix, this.nodeType);
        }

        renumberDataOutputs(this, prefix);

        const dataCount = countDataOutputs(this);
        const primitivesLabel = slotLabelForLinkType(this.nodeType);
        for (let i = 0; i < dataCount; i++) {
            const output = this.outputs[i];
            output.type = this.nodeType;
            output.label = primitivesLabel ?? output.name;
        }

        const indexPos = getIndexOutputSlotIndex(this);
        const indexOutput = this.outputs[indexPos];
        indexOutput.type = "INT";
        indexOutput.name = INDEX_OUTPUT_NAME;
        indexOutput.label = INDEX_OUTPUT_NAME;

        // any_* 出力は末尾に空スロットを1本維持するため、
        // select_index の最大値は (データ出力数 - 2) にする。
        syncSelectIndexWidget(this, (node) => countDataOutputs(node) - 2);
        propagatePrimitiveSplitSync(this);
        this.graph?.setDirtyCanvas?.(true, false);
    };
}

app.registerExtension({
    name: "comfyui-misc.AnyOutputSwitch",
    async beforeRegisterNodeDef(nodeType, nodeData) {
        if (nodeData.name !== NODE_CLASS) {
            return;
        }

        // 出力を2個固定（*, INT）にすると可変出力と index が壊れるため、データ出力は1個だけ定義する
        nodeData.input = nodeData.input ?? {};
        nodeData.input.required = {
            input: ["*"],
            select_index: ["INT", { default: -1, min: -1, max: 4096 }],
            ...(nodeData.input.required ?? {}),
        };
        nodeData.output = ["*"];
        nodeData.output_is_list = [false];
        nodeData.output_name = ["any_01"];

        setupMiscAnyOutputSwitch(nodeType);
    },
});

File: py/any_output_switch.py
"""
Any Output Switch (misc)

入力1本を、複数の出力 any_01, any_02 … のうち
「最初に配線されている出力」だけに渡します。
それ以外の出力先には ExecutionBlocker を返し、下流ノードは実行されません。

最後の index 出力には、選ばれたデータ出力の番号（0始まりの整数）を渡します。
"""

from comfy_execution.graph import ExecutionBlocker

from .utils import ByPassNamesTuple, ByPassTypeTuple, any_type, get_category, get_name

MIN_DATA_OUTPUTS = 1
# web/any_output_switch.js stabilize が最後の接続 data の次に確保する空きスロット数
_STABILIZE_TRAILING_EMPTY_DATA_SLOTS = 1


def _linked_output_slots(prompt, unique_id):
    """
    ワークフロー JSON から、このノード出力に接続されているスロット番号を集める。

    ComfyUI の prompt では、各ノード入力が [元ノードID, 元スロット番号] で保存される。
    """
    uid = str(unique_id)
    slots = []
    for node in prompt.values():
        for value in node.get("inputs", {}).values():
            if isinstance(value, list) and len(value) == 2 and str(value[0]) == uid:
                slots.append(int(value[1]))
    return slots


def _pad_num_data_for_stabilize(num_data, max_slot):
    """index が data スロットに被らないよう、stabilize 分の空き data を num_data に含める。"""
    return max(num_data, max_slot + _STABILIZE_TRAILING_EMPTY_DATA_SLOTS + 1)


def _compute_output_layout(linked_slots, min_data_outputs):
    """
    データ出力の本数と index 出力位置を決める。

    返却タプルは result[0..num_data-1] が data、result[num_data] が index (int)。
    """
    if not linked_slots:
        num_data = min_data_outputs
        return [], num_data, num_data

    sorted_slots = sorted(set(int(s) for s in linked_slots))
    max_slot = sorted_slots[-1]

    if len(sorted_slots) == 1 and sorted_slots[0] == min_data_outputs:
        num_data = min_data_outputs
        return [], num_data, num_data

    contiguous_from_zero = sorted_slots == list(range(max_slot + 1))
    if contiguous_from_zero:
        num_data = max(min_data_outputs, max_slot + 1)
        data_slots = sorted_slots
    else:
        num_data = max(min_data_outputs, max_slot)
        data_slots = [s for s in sorted_slots if s < num_data]

    num_data = _pad_num_data_for_stabilize(num_data, max_slot)
    return data_slots, num_data, num_data


class MiscAnyOutputSwitch:
    """1入力・可変複数出力のスイッチ（Python 側）。"""

    NAME = get_name("Any Output Switch")
    CATEGORY = get_category()

    @classmethod
    def INPUT_TYPES(cls):  # pylint: disable=invalid-name
        return {
            "required": {
                "input": (any_type,),
                "select_index": ("INT", {"default": -1, "min": -1, "max": 4096}),
            },
            "hidden": {
                "prompt": "PROMPT",
                "unique_id": "UNIQUE_ID",
            },
        }

    RETURN_TYPES = ByPassTypeTuple((any_type,))
    RETURN_NAMES = ByPassNamesTuple(("any_01",))
    FUNCTION = "switch"

    def switch(self, input, select_index=-1, prompt=None, unique_id=None, **kwargs):
        """
        入力を 1 本だけ有効化して返す。

        - 選ばれた data 出力: input
        - それ以外の data 出力: ExecutionBlocker
        - 末尾: 選ばれた index (0 始まり)
        """
        linked = (
            _linked_output_slots(prompt, unique_id)
            if prompt is not None and unique_id is not None
            else []
        )

        data_slots, num_data, _index_slot = _compute_output_layout(linked, MIN_DATA_OUTPUTS)

        if select_index is not None and select_index >= 0:
            selected_index = select_index
            num_data = max(num_data, selected_index + 1, MIN_DATA_OUTPUTS)
        else:
            selected_index = min(data_slots) if data_slots else -1

        result = []
        for i in range(num_data):
            if i == selected_index:
                result.append(input)
            else:
                result.append(ExecutionBlocker(None))

        result.append(int(selected_index))
        return tuple(result)


NODE_CLASS_MAPPINGS = {
    MiscAnyOutputSwitch.NAME: MiscAnyOutputSwitch,
}

NODE_DISPLAY_NAME_MAPPINGS = {
    MiscAnyOutputSwitch.NAME: MiscAnyOutputSwitch.NAME,
}
